src/protocols/secure_channel/CheckinMessage.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements the utility class for the Matter Checkin protocol.
+ */
+
+#include "CheckinMessage.h"
+#include <lib/core/CHIPCore.h>
+
+namespace chip {
+namespace Protocols {
+namespace SecureChannel {
+
+static const uint16_t mMaxPlainTextSize = CHIP_CHECK_IN_APP_DATA_MAX_SIZE + CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES;
+static const uint16_t mMaxWorBufferSize = mMaxPlainTextSize + sizeof(uint32_t);
+static const uint16_t minPayloadSize = CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES + sizeof(uint32_t) + CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES;
+
+CHIP_ERROR CheckinMessage::GenerateCheckingMessagePayload(Crypto::Aes128KeyHandle & key, uint32_t counter, ByteSpan appData,
+                                                          uint8_t * output, uint16_t * payloadSize)
+{
+    VerifyOrReturnError(nullptr != output, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(appData.size() <= CHIP_CHECK_IN_APP_DATA_MAX_SIZE, CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint8_t * outputPtr = output;
+
+    {
+        chip::Crypto::HMAC_sha shaHandler;
+        uint8_t nonceWorkBuffer[CHIP_CRYPTO_HASH_LEN_BYTES] = { 0 };
+        ReturnErrorOnFailure(shaHandler.HMAC_SHA256(key.As<Aes128KeyByteArray>(), sizeof(Aes128KeyByteArray) * 8,
+                                                    (uint8_t *) &counter, sizeof(counter), nonceWorkBuffer,
+                                                    CHIP_CRYPTO_HASH_LEN_BYTES));
+        memcpy(outputPtr, nonceWorkBuffer, CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES);
+        outputPtr += CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES;
+    }
+
+    // Concatenate the App data (if any) with the counter value
+    uint8_t workbuffer[mMaxWorBufferSize];
+    uint8_t mic[CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES];
+    uint8_t * p = workbuffer;
+    Encoding::LittleEndian::Write32(p, counter);
+    memcpy(workbuffer + sizeof(counter), appData.data(), appData.size());
+
+    err = Crypto::AES_CCM_encrypt(workbuffer, appData.size() + sizeof(counter), nullptr, 0, key, output,
+                                  CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES, outputPtr, mic, CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES);
+
+    memcpy(outputPtr + appData.size() + sizeof(counter), mic, CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES);
```

Comment:
Shouldn't we return immediately if err is a failure here?

Reply:
fixed